src/service/broker.ts

```diff
@@ -168,6 +233,50 @@ export class OrderBroker implements Interfaces.IOrderBroker {
         this.onOrderUpdate(rpt);
     };
 
+    private _reTrade = (reTrades: Models.Trade[], trade: Models.Trade) => {
+      var gowhile = true;
+      while (gowhile && trade.quantity>0 && reTrades!=null && reTrades.length) {
+        var reTrade = reTrades.shift();
+        gowhile = false;
+        for(var i = 0;i<this._trades.length;i++) {
+          if (this._trades[i].tradeId==reTrade.tradeId) {
+            gowhile = true;
+            var allocQty = Math.min(trade.quantity, this._trades[i].quantity - this._trades[i].alloc);
+            this._trades[i].time = trade.time;
+            this._trades[i].allocprice = ((allocQty*trade.price) + (this._trades[i].alloc*this._trades[i].allocprice)) / (this._trades[i].alloc+allocQty);
+            this._trades[i].alloc += allocQty;
+            trade.quantity -= allocQty;
+            trade.value = Math.abs(trade.price*trade.quantity);
+            if (this._trades[i].quantity<=this._trades[i].alloc)
+              this._trades[i].value = Math.abs((this._trades[i].quantity*this._trades[i].price)-(this._trades[i].alloc*this._trades[i].allocprice));
```

Comment:
Similar to the discussion of IPersist, this code seems fairly out of place and blends quoting logic-specific code in a place that shouldn't really care about quoting mode.

There are several IOrderEntryGateways providing exchange-specific implementations of order entry but translates their data into a common API. The OrderBroker's single responsibility is to provide a stateful abstraction around an IOrderEntryGateway, so users of order entry can query state, listen to order updates, and submit orders/cancels. It should be entirely possible to take the OrderBroker implementation as is and drop it into an arbitrage bot or TA bot (I've actually done this several times while testing new trading styles). This logic doesn't belong at this level and should be at a level above the OrderBroker, abstracted out.